prepare_stmt.go

```diff
@@ -48,12 +46,18 @@ func (db *PreparedStmtDB) Close() {
 	db.Mux.Lock()
 	defer db.Mux.Unlock()
 
-	for _, query := range db.PreparedSQL {
-		if stmt, ok := db.Stmts[query]; ok {
-			delete(db.Stmts, query)
-			go stmt.Close()
-		}
+	for _, stmt := range db.Stmts {
+		go func(s *Stmt) {
+			// make sure the stmt must finish preparation first
+			<-s.prepared
```

Comment:
Can you describe what happens if we don't wait for `s.prepared`?

Reply:
while we prepare a new statement, the new Stmt register itself to the db.Stmt map (with sql.Stmt* field value nil), and unlock the mutex at line 100:
https://github.com/go-gorm/gorm/blob/4a50b36f638c6899089e6e3457425528ce693933/prepare_stmt.go#L98-L124

If we call Close function at this time(after line 101 and before line 119), don't wait for `s.prepared`, the condition of s.Stmt(the *sql.Stmt field) is unknown, it could be:

### 1) a nil pointer, and the preparation step not done yet(line 110), and successfully prepared after we call Close

in this scenario, we must not call Close of the Stmt as it is a null pointer, which result in a Stmt leak(the prepared Stmt generated by line 110 never called Close after that)

### 2) a nil pointer, and the preparation step not done yet(line 110), but failed to prepare after we call Close

in this scenario, we must not call Close of the Stmt as it is a null pointer too, but it is ok

So waiting for `s.prepared` to avoid Stmt leak.